File: python/agentpin/pinning.py
"""TOFU (Trust On First Use) key pinning store for AgentPin."""

import json
from datetime import datetime, timezone
from typing import Dict, List, Optional

from .jwk import jwk_thumbprint
from .types import AgentPinError, ErrorCode, TrustLevel


class PinningResult:
    FIRST_USE = "first_use"
    MATCHED = "matched"
    CHANGED = "changed"


class KeyPinStore:
    """In-memory TOFU key pinning store."""

    def __init__(self) -> None:
        self.domains: Dict[str, dict] = {}

    def check_and_pin(self, domain: str, jwk: dict) -> str:
        """Check a key against the pin store. TOFU pins on first use.

        Returns:
            One of PinningResult values.
        """
        key_hash = jwk_thumbprint(jwk)
        now = datetime.now(timezone.utc).isoformat()

        if domain in self.domains:
            pinned = self.domains[domain]
            for pk in pinned["pinned_keys"]:
                if pk["public_key_hash"] == key_hash:
                    pk["last_seen"] = now
                    return PinningResult.MATCHED
            return PinningResult.CHANGED

        # First time — TOFU pin
        self.domains[domain] = {
            "domain": domain,
            "pinned_keys": [
                {
                    "kid": jwk.get("kid", ""),
                    "public_key_hash": key_hash,
                    "first_seen": now,
                    "last_seen": now,
                    "trust_level": TrustLevel.TOFU,
                }
            ],
        }
        return PinningResult.FIRST_USE

    def add_key(self, domain: str, jwk: dict) -> None:
        """Add a key to an existing domain's pin set (e.g., during key rotation)."""
        key_hash = jwk_thumbprint(jwk)
        now = datetime.now(timezone.utc).isoformat()

        if domain not in self.domains:
            self.domains[domain] = {"domain": domain, "pinned_keys": []}

        pinned = self.domains[domain]
        if not any(pk["public_key_hash"] == key_hash for pk in pinned["pinned_keys"]):
            pinned["pinned_keys"].append(
                {
                    "kid": jwk.get("kid", ""),
                    "public_key_hash": key_hash,
                    "first_seen": now,
                    "last_seen": now,
                    "trust_level": TrustLevel.TOFU,
                }
            )

    def get_domain(self, domain: str) -> Optional[dict]:
        """Get pinned domain info."""
        return self.domains.get(domain)

    def to_json(self) -> str:
        """Serialize the store to JSON."""
        return json.dumps(list(self.domains.values()), indent=2)

    @classmethod
    def from_json(cls, json_str: str) -> "KeyPinStore":
        """Deserialize the store from JSON."""
        store = cls()
        domains: List[dict] = json.loads(json_str)
        for d in domains:
            store.domains[d["domain"]] = d
        return store


def check_pinning(store: KeyPinStore, domain: str, jwk: dict) -> str:
    """Check pinning and raise an error if key has changed.

    Returns:
        PinningResult value.

    Raises:
        AgentPinError: if key has changed since last pinned.
    """
    result = store.check_and_pin(domain, jwk)
    if result == PinningResult.CHANGED:
        raise AgentPinError(
            ErrorCode.KEY_PIN_MISMATCH,
            f"Key for domain '{domain}' has changed since last pinned (kid: '{jwk.get('kid', '')}')",
        )
    return result

File: go/pkg/pinning/pinning.go
// Package pinning provides a TOFU (trust-on-first-use) key pin store keyed
// by domain. The store persists by JSON serialization and is used by the
// AgentPin verifier to detect key changes.
package pinning

import (
	"encoding/json"
	"fmt"
	"sync"
	"time"

	"github.com/ThirdKeyAi/agentpin/go/pkg/jwk"
	"github.com/ThirdKeyAi/agentpin/go/pkg/types"
)

// Result is the outcome of a CheckAndPin call.
type Result int

const (
	// ResultFirstUse indicates the domain was unknown and the key is now pinned.
	ResultFirstUse Result = iota
	// ResultMatched indicates the key matches a previously pinned key.
	ResultMatched
	// ResultChanged indicates the domain was known but the key does not match
	// any pinned key — likely key rotation or attack.
	ResultChanged
)

// String returns a stable label for r.
func (r Result) String() string {
	switch r {
	case ResultFirstUse:
		return "first_use"
	case ResultMatched:
		return "pinned"
	case ResultChanged:
		return "changed"
	}
	return "unknown"
}

// KeyPinStore is a concurrent-safe in-memory TOFU pin store.
type KeyPinStore struct {
	mu      sync.Mutex
	domains map[string]*types.PinnedDomain
}

// NewKeyPinStore creates a new empty store.
func NewKeyPinStore() *KeyPinStore {
	return &KeyPinStore{domains: make(map[string]*types.PinnedDomain)}
}

// CheckAndPin checks j against the pinned keys for domain. On first use it
// pins j and returns ResultFirstUse.
func (s *KeyPinStore) CheckAndPin(domain string, j *types.JWK) Result {
	s.mu.Lock()
	defer s.mu.Unlock()

	hash := jwk.JWKThumbprint(j)
	now := time.Now().UTC().Format(time.RFC3339)

	if pd, ok := s.domains[domain]; ok {
		for i := range pd.PinnedKeys {
			if pd.PinnedKeys[i].PublicKeyHash == hash {
				pd.PinnedKeys[i].LastSeen = now
				return ResultMatched
			}
		}
		return ResultChanged
	}

	s.domains[domain] = &types.PinnedDomain{
		Domain: domain,
		PinnedKeys: []types.PinnedKey{
			{
				Kid:           j.Kid,
				PublicKeyHash: hash,
				FirstSeen:     now,
				LastSeen:      now,
				TrustLevel:    types.TrustTOFU,
			},
		},
	}
	return ResultFirstUse
}

// AddKey adds an additional pinned key to a domain (e.g., during key rotation).
func (s *KeyPinStore) AddKey(domain string, j *types.JWK) {
	s.mu.Lock()
	defer s.mu.Unlock()

	hash := jwk.JWKThumbprint(j)
	now := time.Now().UTC().Format(time.RFC3339)

	pd, ok := s.domains[domain]
	if !ok {
		pd = &types.PinnedDomain{Domain: domain}
		s.domains[domain] = pd
	}
	for _, k := range pd.PinnedKeys {
		if k.PublicKeyHash == hash {
			return
		}
	}
	pd.PinnedKeys = append(pd.PinnedKeys, types.PinnedKey{
		Kid:           j.Kid,
		PublicKeyHash: hash,
		FirstSeen:     now,
		LastSeen:      now,
		TrustLevel:    types.TrustTOFU,
	})
}

// GetDomain returns the pinned-domain entry for domain, or nil.
func (s *KeyPinStore) GetDomain(domain string) *types.PinnedDomain {
	s.mu.Lock()
	defer s.mu.Unlock()
	pd, ok := s.domains[domain]
	if !ok {
		return nil
	}
	cp := *pd
	cp.PinnedKeys = append([]types.PinnedKey(nil), pd.PinnedKeys...)
	return &cp
}

// MarshalJSON serializes the store as a JSON array of PinnedDomain (matching
// the Rust port's `to_json` shape).
func (s *KeyPinStore) MarshalJSON() ([]byte, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	domains := make([]*types.PinnedDomain, 0, len(s.domains))
	for _, d := range s.domains {
		domains = append(domains, d)
	}
	return json.MarshalIndent(domains, "", "  ")
}

// LoadFromJSON populates s from the JSON array produced by MarshalJSON.
func (s *KeyPinStore) LoadFromJSON(data []byte) error {
	var domains []types.PinnedDomain
	if err := json.Unmarshal(data, &domains); err != nil {
		return err
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	s.domains = make(map[string]*types.PinnedDomain, len(domains))
	for i := range domains {
		d := domains[i]
		s.domains[d.Domain] = &d
	}
	return nil
}

// CheckPinning runs CheckAndPin and returns ResultChanged as a typed
// VerificationError so verifiers can fail closed cleanly.
func CheckPinning(store *KeyPinStore, domain string, j *types.JWK) (Result, error) {
	r := store.CheckAndPin(domain, j)
	if r == ResultChanged {
		return r, types.NewVerificationError(
			types.ErrKeyPinMismatch,
			fmt.Sprintf("Key for domain '%s' has changed since last pinned (kid: '%s')", domain, j.Kid),
		)
	}
	return r, nil
}
